File: tests/fixtures/langs/python/widgets.py
"""Python extraction fixture: store, service, controller, one route."""

# Größe — 寸法 📦: 2-, 3- and 4-byte UTF-8 before every node, so a byte offset
# used on decoded text would shift every name, line and doc below this point.
LABEL = "寸法 📦 größe"


class WidgetStore:
    """Persists widgets."""

    def save(self, item):
        """Store one widget — größe 寸法 📦."""
        return item

    def größe(self):
        """Größe of the store — 寸法."""
        return 0


def grade(score, bonus):
    """Hand-counted: complexity 6 (1 + for, if, and, if, elif), depth 2, 2 params."""
    total = score
    for b in bonus:
        if b > 0 and total < 100:
            total += b
    if total > 90:
        return "A"
    elif total > 50:
        return "B"
    return "C"


class WidgetService:
    """Applies widget rules, then persists."""

    def __init__(self, store: WidgetStore):
        self.store = store

    def place(self, item):
        """Place one widget through the store."""
        return self.store.save(item)


class WidgetController:
    """HTTP entry point for widgets."""

    def __init__(self, service: WidgetService):
        self.service = service

    @router.post("/widgets")
    def create(self, item):
        """POST /widgets."""
        return self.service.place(item)

File: tests/fixtures/langs/swift/Widgets.swift
// Größe — 寸法 📦: 2-, 3- and 4-byte UTF-8 before every node, so a byte offset
// used on decoded text would shift every name, line and doc below this point.
let label = "寸法 📦 größe"

/// Persists widgets.
class WidgetStore {
    /// Store one widget — größe 寸法 📦.
    func save(_ item: String) -> String {
        return item
    }

    /// Größe of the store — 寸法.
    func größe() -> Int {
        return 0
    }
}

/// Applies widget rules, then persists.
class WidgetService {
    let store: WidgetStore

    init(store: WidgetStore) {
        self.store = store
    }

    /// Place one widget through the store.
    func place(_ item: String) -> String {
        return store.save(item)
    }

    /// Hand-counted: complexity 6 (1 + for, if, &&, if, else if), depth 2, 2 params.
    func grade(score: Int, bonus: [Int]) -> String {
        var total = score
        for b in bonus {
            if b > 0 && total < 100 {
                total += b
            }
        }
        if total > 90 {
            return "A"
        } else if total > 50 {
            return "B"
        }
        return "C"
    }
}
